### PR TITLE
Add ability to configure component-preload and custom bundles

### PR DESCRIPTION
**Pull request by tommyvinhlam**

### CODE REVIEW

File: lib/types/application/ApplicationBuilder.js [hunk 1]
```diff
@@ -82,16 +84,23 @@ class ApplicationBuilder extends AbstractBuilder {
 			});
 		});
 
-		this.addTask("generateAppPreload", () => {
+		var aComponentPreload = project.builder.componentPreload;
```
Comment (RandomByte) [inline lib/types/application/ApplicationBuilder.js]: No hungarian notation to be consistent with the general style applied here.

File: lib/types/application/ApplicationBuilder.js [hunk 2]
```diff
@@ -104,6 +113,25 @@ class ApplicationBuilder extends AbstractBuilder {
 			});
 		});
 
+		debugger;
```
Comment (RandomByte) [inline lib/types/application/ApplicationBuilder.js]: debugger statement

File: lib/types/application/ApplicationBuilder.js (hunk 1 above)
Comment (tommyvinhlam) [reply lib/types/application/ApplicationBuilder.js]: done

File: lib/types/application/ApplicationBuilder.js (hunk 2 above)
Comment (tommyvinhlam) [reply lib/types/application/ApplicationBuilder.js]: done

**Comment by tommyvinhlam** [comment]
The test for the component-preload generation still fails, as the output of the root component-preload still contains the sub components

File: lib/types/application/ApplicationBuilder.js [hunk 3]
```diff
@@ -37,7 +39,7 @@ class ApplicationBuilder extends AbstractBuilder {
 				options: {
 					copyright: project.metadata.copyright,
 					pattern: "/**/*.js"
-				}
+				},
```
Comment (tommyvinhlam) [inline lib/types/application/ApplicationBuilder.js]: This comma should be removed

File: test/expected/build/application.g/dest/Component-dbg.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+sap.ui.define(["sap/ui/core/UIComponent"], function(UIComponent){
+	"use strict";
+	return UIComponent.extend('application.g.Component', {
+		metadata: {
+			manifest: "json"
+		}
+	});
+});
```
Comment (tommyvinhlam) [inline test/expected/build/application.g/dest/Component-dbg.js]: Add new line to fix eslint issue

File: test/expected/build/application.g/dest/manifest.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+        "_version": "1.1.0",
+        "id": "application.g",
+        "type": "application",
+        "applicationVersion": {
+            "version": "1.2.2"
+        },
+        "embeds": ["embedded"],
+        "title": "{{title}}"
+    }
+}
```
Comment (tommyvinhlam) [inline test/expected/build/application.g/dest/manifest.json]: Add new line

File: test/expected/build/application.g/dest/subcomponentA/Component-dbg.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+sap.ui.define(["sap/ui/core/UIComponent"], function(UIComponent){
+	"use strict";
+	return UIComponent.extend('application.g.subcomponentA.Component', {
+		metadata: {
+			manifest: "json"
+		}
+	});
+});
```
Comment (tommyvinhlam) [inline test/expected/build/application.g/dest/subcomponentA/Component-dbg.js]: Add new line

**Comment by RandomByte** [comment]
Open points:
- explicit configuration for library preloads
- Bundle generation for sources located within `/test-resources/`

**Pull request merged**